enhancements/microshift/microshift-telemetry.md

```diff
@@ -0,0 +1,299 @@
+Introduce an automatic, opt-out mechanism for connected MicroShift clusters to
+report their status to Red Hat. Each MicroShift cluster should send the
+following metrics (at least) to the Telemetry API:
+* Number of available cores/RAM/disk
+* Average utilization of cores/RAM/disk (% of total)
+* Number of namespaces
+* Number of running pods
+* Number of container images on disk
+* Number of routes/ingress/services
+* Number of CRDs
+* OS version / Type (rpm / ostree)
+
+These metrics should be sent at least once a day.
+
+### Metrics details
+Based on [Telemetry API](https://rhobs-handbook.netlify.app/products/openshiftmonitoring/telemetry.md/)
+MicroShift will send metrics using Prometheus format. There are several ways
+of using the API: [using the client to forward metrics](https://github.com/openshift/telemeter/tree/main?tab=readme-ov-file#upload-endpoint-receive-metrics-in-client_modelmetricfamily-format-from-telemeter-client-currently-used-by-cmo),
+or [direct write requests](https://github.com/openshift/telemeter/tree/main?tab=readme-ov-file#metricsv1receive-endpoint-receive-metrics-in-prompbwriterequest-format-from-any-client).
+Using the client is not an option because that requires a local Prometheus
+instance, not viable in the typical resource constrained MicroShift deployment.
+The direct write requests take raw Prometheus data and require both special
+labels (`_id` for the cluster id) and crafting the authentication headers
+(cluster id and pull secret in a specific format in HTTP headers).
+
+OpenShift is using the same API and the same backend, so we need a way to
+distinguish MicroShift in the pool of metrics. For this we can use labels, as
+it does not require applying for new supported metrics. MicroShift metrics
+labels summary:
+| Label | Values | Approximate max size |
+|:---|:---|:---|
+|_id|cluster id|40B|
+|label_kubernetes_io_arch|amd64, aarch64|32B|
+|resource|Used to specify K8s resource names: pods, namespaces, etc.|20-50B|
+|version|microshift version|16B|
+|ostree_commit|OStree commit id, if the system is deployed using ostree. This helps identifying fleets|80B|
+
+Metrics from MicroShift are already supported in the API because OpenShift is
+using them. List follows:
+* cluster:capacity_cpu_cores:sum. Number of allocated CPUs for MicroShift.
+* cluster:capacity_memory_bytes:sum. Number of bytes of memory allocated for
+  MicroShift.
+* cluster:cpu_usage_cores:sum. Usage of CPU in percentage.
+* cluster:memory_usage_bytes:sum. Usage of memory in percentage.
+* cluster:usage:resources:sum. Usage of k8s resources, in count. Number of
+  pods, namespaces, services, etc.
+* cluster:usage:containers:sum. Number of active containers.
+* cluster_version. Information about installed version.
+* instance:etcd_object_counts:sum. Number of objects etcd contains.
+
+Combining metrics with their labels:
+| Metric | Labels |
+|:---|:---|
+|cluster:capacity_cpu_cores:sum|_id, label_kubernetes_io_arch, ostree_commit|
```

Comment:
What about the other labels found in OCP: label_beta_kubernetes_io_instance_type, label_node_role_kubernetes_io, label_node_openshift_io_os_id? Are there useful values in the case of MicroShift?

Note that cluster:capacity_cpu_cores:sum is eventually used by the subscription watch service to report core usage and we don't to introduce side effects that would disrupt this reporting. From what I can tell, the proposed implementation should be safe but it would be good to mention it in the proposal.

https://github.com/openshift/telemeter/blob/f63fef91d659c4127838da6e3dc696e07c0e5536/jsonnet/telemeter/rules.libsonnet#L182-L198

Reply:
> I understand that OpenShift is always having rhcos instead of linux/rhel as the OS in every deployment (including hypershift, hcp, etc.)?

No  there are different values depending on the node OS (e.g. rhel, linux, windows, ...). Again the goal here should be to emit the same label set: even if the values are always the same for MicroShift instances, data consistency on the Telemetry server side is very important.

Reply:
Since MicroShift is not using `label_node_kubernetes_io_instance_type` yet, we will fix it to `rhde` to signal this cluster belongs to Red Hat Device Edge. This way we have the same labels, same semantics, only another possible value to help us distinguish the product.